tooling/nargo_fmt/src/rewrite/typ.rs

```diff
@@ -0,0 +1,69 @@
+use noirc_frontend::{UnresolvedType, UnresolvedTypeData};
+
+use crate::visitor::{FmtVisitor, Shape};
+
+pub(crate) fn rewrite(visitor: &FmtVisitor, _shape: Shape, typ: UnresolvedType) -> String {
+    match typ.typ {
+        UnresolvedTypeData::Array(length, element) => {
+            let typ = rewrite(visitor, _shape, *element);
+            if let Some(length) = length {
+                let length = visitor.slice(length.span());
+                format!("[{typ}; {length}]")
+            } else {
+                format!("[{typ}]")
+            }
+        }
+        UnresolvedTypeData::Parenthesized(typ) => {
+            let typ = rewrite(visitor, _shape, *typ);
+            format!("({typ})")
+        }
+        UnresolvedTypeData::MutableReference(typ) => {
+            let typ = rewrite(visitor, _shape, *typ);
+            format!("&mut {typ}")
+        }
+        UnresolvedTypeData::Tuple(mut types) => {
+            if types.len() == 1 {
+                let typ = types.pop().unwrap();
+                let typ = rewrite(visitor, _shape, typ);
+
+                format!("({typ},)")
+            } else {
+                let types: Vec<_> =
+                    types.into_iter().map(|typ| rewrite(visitor, _shape, typ)).collect();
+                let types = types.join(", ");
+                format!("({types})")
+            }
+        }
+        UnresolvedTypeData::Function(args, return_type, env) => {
+            let env_slice = visitor.slice(env.span.unwrap());
+
+            let env = if env_slice == "(" {
```

Comment:
This check seems very fragile. Is there a way to avoid it, e.g. checking if the span is empty?

Reply:
Or should we consider implementing something like an Option<Type>?